standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/columnstats/aggr/DateColumnStatsAggregator.java

```diff
@@ -63,7 +63,7 @@ public ColumnStatisticsObj aggregate(List<ColStatsObjWithSourceInfo> colStatsWit
         LOG.trace("doAllPartitionContainStats for column: {} is: {}", colName, doAllPartitionContainStats);
       }
       DateColumnStatsDataInspector dateColumnStats =
-          (DateColumnStatsDataInspector) cso.getStatsData().getDateStats();
+          new DateColumnStatsDataInspector(cso.getStatsData().getDateStats());
```

Comment:
The constructor that has been added will always be called here (and in those calls below), as ```getDataStats``` always returns a ```DataColumnStatsData``` object. Hence we will defeat the purpose of the inspector, which was avoiding unnecessary serialization/deserialization of the bit vectors (check ```DateColumnStatsDataInspector(DateColumnStatsDataInspector other)``` constructor and ```getNdvEstimator``` and related methods to get more context).

We should try to obtain a repro case to understand why we are not getting an inspector instance in some cases. One reason may be the ```SharedCache```, which caches thrift ```DateColumnStatsData``` objects, hence we may need to create inspector objects when we retrieve those objects from the cache (this would not require any change in the aggregator).

Alternatively, if that is not possible, I guess the hackish solution is to check whether ```cso.getStatsData().getDateStats()``` is already an instance of DateColumnStatsDataInspector, and skip calling the constructor in those cases. In addition, we may replace the stats object in the ```ColumnStatisticsObj``` so we do not serialize/deserialize again and again.

I hope that makes sense.

Reply:
The `ColumnStatisticsObj` comes from  [here](https://github.com/apache/hive/blob/3104d4756a6ff4c7853e4a17bbef13317770efd2/ql/src/java/org/apache/hadoop/hive/ql/stats/ColStatsProcessor.java#L189). Since it's using the thrift API I assume only `LongColumnStatsData` can be sent as those are the ones defined in the thrift file (not `LongColumnStatsDataInspector`).
After the RPC call the stats are merged [here](https://github.com/apache/hive/blob/3104d4756a6ff4c7853e4a17bbef13317770efd2/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L7613) (or inside [here](https://github.com/apache/hive/blob/3104d4756a6ff4c7853e4a17bbef13317770efd2/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L7599)) and a simple `LongColumnStatsData` is inside the `ColumnStatisticsObj`. Since `LongColumnStatsMerger.merge` can be called from other places with a `LongColumnStatsDataInspector` I think the best is to assume inside it that both things can arrive and implement the second approach you proposed, wdyt @jcamachor ?